lib/activerecord/delay_touching/state.rb

```diff
@@ -15,6 +15,13 @@ def initialize
       end
 
       def updated(attr, records)
+        # Records may have been changed since they were added to the set.  For instance, if an error
+        # occurred and a Rollback was generated, Rails might change the record's id from an integer
+        # to a nil (if it's a new record).  Since it was stored in the set originally, using the
+        # hash of the id of the record (thanks to Rails' magic), it won't be removed because now the
+        # hash is different and it isn't found in the set.
+        @records[attr] = Set.new(@records[attr]) # recreate the Set so it's reliable
```

Comment:
I can't say it doesn't break anything.  I can't think of anything it would break.  I haven't seen any evidence that it breaks anything.  The difficulty in handling the bad records while subtracting them is that the code is in ruby's Set class.  We could monkey patch it (blech), or rewrite the gem to use something else, but then we'd have to replace the uniqueness quality with something.

The bottom line is, once the object in the Set has been changed, the set becomes unreliable (this is mentioned in the documentation for Set).  That unreliable-ness makes it difficult to fix the set.  It's possible I could do something like:
```ruby
@records[attr].delete_if {|e| e.id.nil?}
```
Maybe that would perform better.

Do you have a specific technique in mind?